services/s3/src/main/java/software/amazon/awssdk/services/s3/parsing/S3Uri.java

```diff
@@ -0,0 +1,262 @@
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        S3Uri s3Uri = (S3Uri) o;
+        return Objects.equals(uri, s3Uri.uri)
+               && Objects.equals(bucket, s3Uri.bucket)
+               && Objects.equals(key, s3Uri.key)
+               && Objects.equals(region, s3Uri.region)
+               && Objects.equals(isPathStyle, s3Uri.isPathStyle)
+               && Objects.equals(queryParams, s3Uri.queryParams);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = uri != null ? uri.hashCode() : 0;
+        result = 31 * result + (bucket != null ? bucket.hashCode() : 0);
+        result = 31 * result + (key != null ? key.hashCode() : 0);
+        result = 31 * result + (region != null ? region.hashCode() : 0);
+        result = 31 * result +  Boolean.hashCode(isPathStyle);
+        result = 31 * result + (queryParams != null ? queryParams.hashCode() : 0);
+        return result;
+    }
+
+    /**
+     * A builder for creating a {@link S3Uri}
+     */
+    public static final class Builder implements CopyableBuilder<Builder, S3Uri> {
+        private URI uri;
+        private String bucket;
+        private String key;
+        private Region region;
+        private boolean isPathStyle;
+        private Map<String, List<String>> queryParams;
+
+        private Builder() {
+        }
+
+        private Builder(S3Uri s3Uri) {
+            this.uri = s3Uri.uri;
+            this.bucket = s3Uri.bucket;
+            this.key = s3Uri.key;
+            this.region = s3Uri.region;
+            this.isPathStyle = s3Uri.isPathStyle;
+            this.queryParams = s3Uri.queryParams;
+        }
+
+        /**
+         * Configure the URI
+         */
+        public Builder uri(URI uri) {
+            this.uri = uri;
+            return this;
+        }
+
+        /**
+         * Configure the bucket
+         */
+        public Builder bucket(String bucket) {
+            this.bucket = bucket;
+            return this;
+        }
+
+        /**
+         * Configure the key
+         */
+        public Builder key(String key) {
+            this.key = key;
+            return this;
+        }
+
+        /**
+         * Configure the region
+         */
+        public Builder region(Region region) {
+            this.region = region;
+            return this;
+        }
+
+        /**
+         * Configure the path style flag
+         */
+        public Builder isPathStyle(boolean isPathStyle) {
+            this.isPathStyle = isPathStyle;
+            return this;
+        }
+
+        /**
+         * Configure the map of query parameters
+         */
+        public Builder queryParams(Map<String, List<String>> queryParams) {
+            this.queryParams = queryParams;
+            return this;
+        }
+
+        /**
+         * Return the configured URI
+         */
+        public URI uri() {
```

Comment:
Seems a bit odd that we just expose getter for one of the property. Can we just pass uri to `addQueryParamsIfNeeded(S3Uri.Builder builder, URI uri)` and `parsePathStyleUri` and `parseVirtualHostedStyleUri`?

Reply:
yup makes sense, updated